// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/kpi/kpi-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KPICalculationModels = void 0;
var KPICalculationModels;
(function (KPICalculationModels) {
    /**
     *
     * @export
     * @class RequiredError
     * @extends {Error}
     */
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    KPICalculationModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace ControlSystemEvent
     */
    let ControlSystemEvent;
    (function (ControlSystemEvent) {
        /**
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum[TypeEnum["SHUTDOWN"] = "SHUTDOWN"] = "SHUTDOWN";
            TypeEnum[TypeEnum["NORMALSTOP"] = "NORMAL_STOP"] = "NORMALSTOP";
        })(TypeEnum = ControlSystemEvent.TypeEnum || (ControlSystemEvent.TypeEnum = {}));
    })(ControlSystemEvent = KPICalculationModels.ControlSystemEvent || (KPICalculationModels.ControlSystemEvent = {}));
})(KPICalculationModels = exports.KPICalculationModels || (exports.KPICalculationModels = {}));
