// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/common/mindsphere-sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MindSphereSdk = void 0;
const advanced_token_1 = require("../advanced-token/advanced-token");
const agent_management_1 = require("../agent/agent-management");
const anomaly_detection_1 = require("../anomaly-detection/anomaly-detection");
const asset_management_1 = require("../asset/asset-management");
const commanding_1 = require("../commanding/commanding");
const data_exchange_1 = require("../data-exchange/data-exchange");
const data_lake_1 = require("../data-lake/data-lake");
const eventanalytics_1 = require("../event-analytics/eventanalytics");
const event_management_1 = require("../event/event-management");
const identity_1 = require("../identity/identity");
const iot_timeseries_1 = require("../iot/iot-timeseries");
const iot_timeseries_aggregate_v4_1 = require("../iotaggregate-v4/iot-timeseries-aggregate-v4");
const iot_timeseries_aggregate_1 = require("../iotaggregate/iot-timeseries-aggregate");
const iot_timeseries_bulk_1 = require("../iotbulk/iot-timeseries-bulk");
const iot_file_1 = require("../iotfile/iot-file");
const jobmanager_1 = require("../jobmanager/jobmanager");
const kpi_1 = require("../kpi/kpi");
const mcapi_1 = require("../mcapi/mcapi");
const messagebroker_1 = require("../messagebroker/messagebroker");
const model_management_1 = require("../model/model-management");
const notification_v4_1 = require("../notification-v4/notification-v4");
const open_edge_1 = require("../open-edge/open-edge");
const policy_1 = require("../policy/policy");
const sdi_v4_1 = require("../sdi/sdi-v4");
const signal_calculation_1 = require("../signal-calculation/signal-calculation");
const signal_validation_1 = require("../signal-validation/signal-validation");
const spectrum_analysis_1 = require("../spectrum/spectrum-analysis");
const tenant_management_1 = require("../tenant/tenant-management");
const trend_prediction_1 = require("../trend/trend-prediction");
const utc_1 = require("../utc/utc");
const sdk_client_1 = require("./sdk-client");
/**
 *
 * MindSphere typescript SDK
 *
 * Runs in browser and in NodeJs.
 *
 * The SDK uses all 4 types of MindSphere Credentials
 * (Agent Credentials, User Credentials, App Credentials,ServiceCredentials).
 *
 * @export
 * @class MindSphereSdk
 */
class MindSphereSdk extends sdk_client_1.SdkClient {
    /**
     * * Asset Management
     *
     * @returns {AssetManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetAssetManagementClient() {
        this._assetManagementClient = this._assetManagementClient || new asset_management_1.AssetManagementClient(this._authenticator);
        return this._assetManagementClient;
    }
    /**
     * * Agent Management
     *
     * @returns {AgentManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetAgentManagementClient() {
        this._agentManagementClient = this._agentManagementClient || new agent_management_1.AgentManagementClient(this._authenticator);
        return this._agentManagementClient;
    }
    /**
     * * Anomaly Detection Client
     *
     * @returns {AnomalyDetectionClient}
     *
     * @memberOf MindSphereSdk
     */
    GetAnomalyDetectionClient() {
        this._anomalyDetectionClient = this._anomalyDetectionClient || new anomaly_detection_1.AnomalyDetectionClient(this._authenticator);
        return this._anomalyDetectionClient;
    }
    /**
     * * Iot File
     *
     * @returns {IotFileClient}
     *
     * @memberOf MindSphereSdk
     */
    GetIoTFileClient() {
        this._iotFileClient = this._iotFileClient || new iot_file_1.IotFileClient(this._authenticator);
        return this._iotFileClient;
    }
    /**
     * * Time Series Bulk
     *
     * @returns {TimeSeriesBulkClient}
     *
     * @memberOf MindSphereSdk
     */
    GetTimeSeriesBulkClient() {
        this._tsBulkClient = this._tsBulkClient || new iot_timeseries_bulk_1.TimeSeriesBulkClient(this._authenticator);
        return this._tsBulkClient;
    }
    /**
     * * Time Series
     *
     * @returns {TimeSeriesClient}
     *
     * @memberOf MindSphereSdk
     */
    GetTimeSeriesClient() {
        this._timeSeriesClient = this._timeSeriesClient || new iot_timeseries_1.TimeSeriesClient(this._authenticator);
        return this._timeSeriesClient;
    }
    /**
     * * Time Series Aggregates
     *
     * @returns {TimeSeriesAggregateClient}
     *
     * @memberOf MindSphereSdk
     */
    GetTimeSeriesAggregateClient() {
        this._timeSeriesAggregateClient =
            this._timeSeriesAggregateClient || new iot_timeseries_aggregate_1.TimeSeriesAggregateClient(this._authenticator);
        return this._timeSeriesAggregateClient;
    }
    /**
     * * Time Series Aggregates
     * * V4 Client
     *
     * @returns {TimeSeriesAggregateClientV4}
     *
     * @memberOf MindSphereSdk
     */
    GetTimeSeriesAggregateClientV4() {
        this._timeSeriesAggregateClientV4 =
            this._timeSeriesAggregateClientV4 || new iot_timeseries_aggregate_v4_1.TimeSeriesAggregateClientV4(this._authenticator);
        return this._timeSeriesAggregateClientV4;
    }
    /**
     * * Event Management
     *
     * @returns {EventManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetEventManagementClient() {
        this._eventManagementClient = this._eventManagementClient || new event_management_1.EventManagementClient(this._authenticator);
        return this._eventManagementClient;
    }
    /**
     *  * Spectrum Analysis Client
     *
     * @returns {SpectrumAnalysisClient}
     *
     * @memberOf MindSphereSdk
     */
    GetSpectrumAnalysisClient() {
        this._spectrumAnalysisClient = this._spectrumAnalysisClient || new spectrum_analysis_1.SpectrumAnalysisClient(this._authenticator);
        return this._spectrumAnalysisClient;
    }
    /**
     * * Signal Validation Client
     *
     * @returns {SignalValidationClient}
     *
     * @memberOf MindSphereSdk
     */
    GetSignalValidationClient() {
        this._signalValidationClient = this._signalValidationClient || new signal_validation_1.SignalValidationClient(this._authenticator);
        return this._signalValidationClient;
    }
    /**
     * * Trend Prediction Client
     *
     * @returns {TrendPredictionClient}
     *
     * @memberOf MindSphereSdk
     */
    GetTrendPredictionClient() {
        this._trendPredictionClient = this._trendPredictionClient || new trend_prediction_1.TrendPredictionClient(this._authenticator);
        return this._trendPredictionClient;
    }
    /**
     * * KPI Calculation Client
     *
     * @returns {KPICalculationClient}
     *
     * @memberOf MindSphereSdk
     */
    GetKPICalculationClient() {
        this._kpiCalculationClient = this._kpiCalculationClient || new kpi_1.KPICalculationClient(this._authenticator);
        return this._kpiCalculationClient;
    }
    /**
     * * MindConnectApiClient
     *
     * @returns {MindConnectApiClient}
     *
     * @memberOf MindSphereSdk
     */
    GetMindConnectApiClient() {
        this._mindConnectApiClient = this._mindConnectApiClient || new mcapi_1.MindConnectApiClient(this._authenticator);
        return this._mindConnectApiClient;
    }
    /**
     * * Identity Management Client
     *
     * @returns {IdentityManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetIdentityManagementClient() {
        this._identityManagementClient =
            this._identityManagementClient || new identity_1.IdentityManagementClient(this._authenticator);
        return this._identityManagementClient;
    }
    /**
     * * Event Analytics Client
     *
     * @returns {EventAnalyticsClient}
     *
     * @memberOf MindSphereSdk
     */
    GetEventAnalyticsClient() {
        this._eventAnalyticsClient = this._eventAnalyticsClient || new eventanalytics_1.EventAnalyticsClient(this._authenticator);
        return this._eventAnalyticsClient;
    }
    /**
     * * Data Lake Client
     *
     * @returns {DataLakeClient}
     *
     * @memberOf MindSphereSdk
     */
    GetDataLakeClient() {
        this._dataLakeClient = this._dataLakeClient || new data_lake_1.DataLakeClient(this._authenticator);
        return this._dataLakeClient;
    }
    /**
     * Usage Transparency Client
     *
     * @returns {UsageTransparencyClient}
     *
     * @memberOf MindSphereSdk
     */
    GetUsageTransparencyClient() {
        this._usageTransparencyClient =
            this._usageTransparencyClient || new utc_1.UsageTransparencyClient(this._authenticator);
        return this._usageTransparencyClient;
    }
    /**
     * * Tenant Management Client
     *
     * @returns {TenantManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetTenantManagementClient() {
        this._tenantManagementClient = this._tenantManagementClient || new tenant_management_1.TenantManagementClient(this._authenticator);
        return this._tenantManagementClient;
    }
    /**
     * * Device Management Client
     *
     * @returns {DeviceManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetDeviceManagementClient() {
        this._deviceManagementClient = this._deviceManagementClient || new open_edge_1.DeviceManagementClient(this._authenticator);
        return this._deviceManagementClient;
    }
    /**
     * * Device Status Management Client
     *
     * @returns {DeviceStatusManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetDeviceStatusManagementClient() {
        this._deviceStatusManagementClient =
            this._deviceStatusManagementClient || new open_edge_1.DeviceStatusManagementClient(this._authenticator);
        return this._deviceStatusManagementClient;
    }
    /**
     * * Device Configuration Management Client
     *
     * @returns {DeviceStatusManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetDeviceConfigurationClient() {
        this._deviceConfigurationClient =
            this._deviceConfigurationClient || new open_edge_1.DeviceConfigurationClient(this._authenticator);
        return this._deviceConfigurationClient;
    }
    /**
     * * Deployment Workflow Client
     *
     * @returns {DeploymentWorkflowClient}
     *
     * @memberOf MindSphereSdk
     */
    GetDeploymentWorkflowClient() {
        this._deploymentWorkflowClient =
            this._deploymentWorkflowClient || new open_edge_1.DeploymentWorkflowClient(this._authenticator);
        return this._deploymentWorkflowClient;
    }
    /**
     * * Edge App Instance Management Client
     *
     * @returns {EdgeAppInstanceManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetEdgeAppInstanceManagementClient() {
        this._edgeAppInstanceManagementClient =
            this._edgeAppInstanceManagementClient || new open_edge_1.EdgeAppInstanceManagementClient(this._authenticator);
        return this._edgeAppInstanceManagementClient;
    }
    /**
     * * Edge AppDeployment Client
     *
     * @returns {EdgeDeploymentClient}
     *
     * @memberOf MindSphereSdk
     */
    GetEdgeDeploymentClient() {
        this._edgeAppDeploymentClient =
            this._edgeAppDeploymentClient || new open_edge_1.EdgeAppDeploymentClient(this._authenticator);
        return this._edgeAppDeploymentClient;
    }
    /**
     * * Edge AppDeployment Client
     *
     * @returns {FirmwareDeploymentClient}
     *
     * @memberOf MindSphereSdk
     */
    GetFirmwareDeploymentClient() {
        this._firmwareAppDeploymentClient =
            this._firmwareAppDeploymentClient || new open_edge_1.FirmwareDeploymentClient(this._authenticator);
        return this._firmwareAppDeploymentClient;
    }
    /**
     * * Model Management Client
     *
     * @returns {ModelManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetModelManagementClient() {
        this._modelManagementClient = this._modelManagementClient || new model_management_1.ModelManagementClient(this._authenticator);
        return this._modelManagementClient;
    }
    /**
     * * Notification Client
     *
     * @returns {NotificationClientV4}
     *
     * @memberOf MindSphereSdk
     */
    GetNotificationClientV4() {
        this._notificationClient = this._notificationClient || new notification_v4_1.NotificationClientV4(this._authenticator);
        return this._notificationClient;
    }
    /**
     * * Job Manager Client
     *
     * @returns {JobManagerClient}
     *
     * @memberOf MindSphereSdk
     */
    GetJobManagerClient() {
        this._jobManagerClient = this._jobManagerClient || new jobmanager_1.JobManagerClient(this._authenticator);
        return this._jobManagerClient;
    }
    /**
     * * Data Exchange Client
     *
     * @returns {DataExchangeClient}
     *
     * @memberOf MindSphereSdk
     */
    GetDataExchangeClient() {
        this._dataExchangeClient = this._dataExchangeClient || new data_exchange_1.DataExchangeClient(this._authenticator);
        return this._dataExchangeClient;
    }
    /**
     * * Signal Calculation Client
     *
     * @returns {SignalCalculationClient}
     *
     * @memberOf MindSphereSdk
     */
    GetSignalCalculationClient() {
        this._signalCalculationClient =
            this._signalCalculationClient || new signal_calculation_1.SignalCalculationClient(this._authenticator);
        return this._signalCalculationClient;
    }
    /**
     * Semantic Data Interconnect Client
     *
     * @returns {SemanticDataInterconnectClient}
     *
     * @memberOf MindSphereSdk
     */
    GetSemanticDataInterConnectClient() {
        this._semanticDataInterConnectClient =
            this._semanticDataInterConnectClient || new sdi_v4_1.SemanticDataInterconnectClient(this._authenticator);
        return this._semanticDataInterConnectClient;
    }
    /**
     * * Message Broker Client
     *
     * @returns {MessageBrokerClient}
     *
     * @memberOf MindSphereSdk
     */
    GetMessageBrokerClient() {
        this._messageBrokerClient = this._messageBrokerClient || new messagebroker_1.MessageBrokerClient(this._authenticator);
        return this._messageBrokerClient;
    }
    /**
     * * Resource Acess Management Client
     *
     * @returns {ResourceAccessManagementClient}
     *
     * @memberOf MindSphereSdk
     */
    GetResourceManagementClient() {
        this._resourceAccessManagementClient =
            this._resourceAccessManagementClient || new policy_1.ResourceAccessManagementClient(this._authenticator);
        return this._resourceAccessManagementClient;
    }
    /**
     * Commanding Client
     *
     * @returns {CommandingClient}
     *
     * @memberOf MindSphereSdk
     */
    GetCommandingClient() {
        this._commandingClient = this._commandingClient || new commanding_1.CommandingClient(this._authenticator);
        return this._commandingClient;
    }
    /**
     * Advanced Token Exchange Client
     *
     * @returns {AdvancedTokenExchangeClient}
     *
     * @memberOf MindSphereSdk
     */
    GetAdvancedTokenExchangeClient() {
        this._advancedTokenExchangeClient =
            this._advancedTokenExchangeClient || new advanced_token_1.AdvancedTokenExchangeClient(this._authenticator);
        return this._advancedTokenExchangeClient;
    }
}
exports.MindSphereSdk = MindSphereSdk;
