// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/event/event-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventManagementModels = void 0;
var EventManagementModels;
(function (EventManagementModels) {
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    EventManagementModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace EventType
     */
    let EventType;
    (function (EventType) {
        /**
         * @export
         * @enum {string}
         */
        let ScopeEnum;
        (function (ScopeEnum) {
            ScopeEnum["LOCAL"] = "LOCAL";
            ScopeEnum["GLOBAL"] = "GLOBAL";
        })(ScopeEnum = EventType.ScopeEnum || (EventType.ScopeEnum = {}));
    })(EventType = EventManagementModels.EventType || (EventManagementModels.EventType = {}));
    /**
     * @export
     * @namespace EventTypePatch
     */
    let EventTypePatch;
    (function (EventTypePatch) {
        /**
         * @export
         * @enum {string}
         */
        let OpEnum;
        (function (OpEnum) {
            OpEnum["Add"] = "add";
            OpEnum["Replace"] = "replace";
        })(OpEnum = EventTypePatch.OpEnum || (EventTypePatch.OpEnum = {}));
    })(EventTypePatch = EventManagementModels.EventTypePatch || (EventManagementModels.EventTypePatch = {}));
    /**
     * @export
     * @namespace EventTypeResponse
     */
    let EventTypeResponse;
    (function (EventTypeResponse) {
        /**
         * @export
         * @enum {string}
         */
        let ScopeEnum;
        (function (ScopeEnum) {
            ScopeEnum["LOCAL"] = "LOCAL";
            ScopeEnum["GLOBAL"] = "GLOBAL";
        })(ScopeEnum = EventTypeResponse.ScopeEnum || (EventTypeResponse.ScopeEnum = {}));
    })(EventTypeResponse = EventManagementModels.EventTypeResponse || (EventManagementModels.EventTypeResponse = {}));
    /**
     * @export
     * @namespace Field
     */
    let Field;
    (function (Field) {
        /**
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum["STRING"] = "STRING";
            TypeEnum["INTEGER"] = "INTEGER";
            TypeEnum["DOUBLE"] = "DOUBLE";
            TypeEnum["BOOLEAN"] = "BOOLEAN";
            TypeEnum["LINK"] = "LINK";
            TypeEnum["TIMESTAMP"] = "TIMESTAMP";
            TypeEnum["UUID"] = "UUID";
            TypeEnum["ENUM"] = "ENUM";
        })(TypeEnum = Field.TypeEnum || (Field.TypeEnum = {}));
    })(Field = EventManagementModels.Field || (EventManagementModels.Field = {}));
    /**
     * @export
     * @namespace JobResource
     */
    let JobResource;
    (function (JobResource) {
        /**
         * @export
         * @enum {string}
         */
        let StateEnum;
        (function (StateEnum) {
            StateEnum["ACCEPTED"] = "ACCEPTED";
            StateEnum["INPROGRESS"] = "IN_PROGRESS";
            StateEnum["FINISHED"] = "FINISHED";
            StateEnum["FINISHEDWITHERROR"] = "FINISHED_WITH_ERROR";
        })(StateEnum = JobResource.StateEnum || (JobResource.StateEnum = {}));
    })(JobResource = EventManagementModels.JobResource || (EventManagementModels.JobResource = {}));
    /**
     * @export
     * @namespace SharingResource
     */
    let SharingResource;
    (function (SharingResource) {
        /**
         * @export
         * @enum {string}
         */
        let ModesEnum;
        (function (ModesEnum) {
            ModesEnum["SHARER"] = "SHARER";
            ModesEnum["RECEIVER"] = "RECEIVER";
            ModesEnum["DEFAULT"] = "DEFAULT";
        })(ModesEnum = SharingResource.ModesEnum || (SharingResource.ModesEnum = {}));
    })(SharingResource = EventManagementModels.SharingResource || (EventManagementModels.SharingResource = {}));
})(EventManagementModels = exports.EventManagementModels || (exports.EventManagementModels = {}));
