// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/anomaly-detection/anomaly-detection-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnomalyDetectionModels = void 0;
var AnomalyDetectionModels;
(function (AnomalyDetectionModels) {
    /**
     *
     * @export
     * @class RequiredError
     * @extends {Error}
     */
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    AnomalyDetectionModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace ReasoningJobInfo
     */
    let ReasoningJobInfo;
    (function (ReasoningJobInfo) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum[StatusEnum["SUBMITTED"] = 'SUBMITTED'] = "SUBMITTED";
            StatusEnum[StatusEnum["RUNNING"] = 'RUNNING'] = "RUNNING";
            StatusEnum[StatusEnum["SUCCEEDED"] = 'SUCCEEDED'] = "SUCCEEDED";
            StatusEnum[StatusEnum["FAILED"] = 'FAILED'] = "FAILED";
        })(StatusEnum = ReasoningJobInfo.StatusEnum || (ReasoningJobInfo.StatusEnum = {}));
    })(ReasoningJobInfo = AnomalyDetectionModels.ReasoningJobInfo || (AnomalyDetectionModels.ReasoningJobInfo = {}));
    /**
     * @export
     * @namespace SubmitTrainingRequest
     */
    let SubmitTrainingRequest;
    (function (SubmitTrainingRequest) {
        /**
         * @export
         * @enum {string}
         */
        let DistanceMeasureAlgorithmEnum;
        (function (DistanceMeasureAlgorithmEnum) {
            DistanceMeasureAlgorithmEnum[DistanceMeasureAlgorithmEnum["EUCLIDEAN"] = 'EUCLIDEAN'] = "EUCLIDEAN";
            DistanceMeasureAlgorithmEnum[DistanceMeasureAlgorithmEnum["MANHATTAN"] = 'MANHATTAN'] = "MANHATTAN";
            DistanceMeasureAlgorithmEnum[DistanceMeasureAlgorithmEnum["CHEBYSHEV"] = 'CHEBYSHEV'] = "CHEBYSHEV";
        })(DistanceMeasureAlgorithmEnum = SubmitTrainingRequest.DistanceMeasureAlgorithmEnum || (SubmitTrainingRequest.DistanceMeasureAlgorithmEnum = {}));
    })(SubmitTrainingRequest = AnomalyDetectionModels.SubmitTrainingRequest || (AnomalyDetectionModels.SubmitTrainingRequest = {}));
    /**
     * @export
     * @namespace TrainingJobInfo
     */
    let TrainingJobInfo;
    (function (TrainingJobInfo) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum[StatusEnum["SUBMITTED"] = 'SUBMITTED'] = "SUBMITTED";
            StatusEnum[StatusEnum["RUNNING"] = 'RUNNING'] = "RUNNING";
            StatusEnum[StatusEnum["SUCCEEDED"] = 'SUCCEEDED'] = "SUCCEEDED";
            StatusEnum[StatusEnum["FAILED"] = 'FAILED'] = "FAILED";
        })(StatusEnum = TrainingJobInfo.StatusEnum || (TrainingJobInfo.StatusEnum = {}));
    })(TrainingJobInfo = AnomalyDetectionModels.TrainingJobInfo || (AnomalyDetectionModels.TrainingJobInfo = {}));
    /**
     *
     * @export
     */
    AnomalyDetectionModels.COLLECTION_FORMATS = {
        csv: ",",
        ssv: " ",
        tsv: "\t",
        pipes: "|",
    };
})(AnomalyDetectionModels = exports.AnomalyDetectionModels || (exports.AnomalyDetectionModels = {}));
