// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/identity/identity-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentityManagementModels = void 0;
var IdentityManagementModels;
(function (IdentityManagementModels) {
    /**
     *
     * @export
     * @class RequiredError
     * @extends {Error}
     */
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    IdentityManagementModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace Group
     */
    let Group;
    (function (Group) {
        /**
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum[TypeEnum["DIRECT"] = "DIRECT"] = "DIRECT";
            TypeEnum[TypeEnum["INDIRECT"] = "INDIRECT"] = "INDIRECT";
        })(TypeEnum = Group.TypeEnum || (Group.TypeEnum = {}));
    })(Group = IdentityManagementModels.Group || (IdentityManagementModels.Group = {}));
    /**
     * @export
     * @namespace ScimGroupMember
     */
    let ScimGroupMember;
    (function (ScimGroupMember) {
        /**
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum[TypeEnum["USER"] = "USER"] = "USER";
            TypeEnum[TypeEnum["GROUP"] = "GROUP"] = "GROUP";
        })(TypeEnum = ScimGroupMember.TypeEnum || (ScimGroupMember.TypeEnum = {}));
    })(ScimGroupMember = IdentityManagementModels.ScimGroupMember || (IdentityManagementModels.ScimGroupMember = {}));
})(IdentityManagementModels = exports.IdentityManagementModels || (exports.IdentityManagementModels = {}));
