// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/jobmanager/jobmanager-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobManagerModels = void 0;
var JobManagerModels;
(function (JobManagerModels) {
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    JobManagerModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace Job
     */
    let Job;
    (function (Job) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum["SUBMITTED"] = "SUBMITTED";
            StatusEnum["STARTING"] = "STARTING";
            StatusEnum["RUNNING"] = "RUNNING";
            StatusEnum["SUCCEEDED"] = "SUCCEEDED";
            StatusEnum["STOPPING"] = "STOPPING";
            StatusEnum["STOPPED"] = "STOPPED";
            StatusEnum["FAILED"] = "FAILED";
        })(StatusEnum = Job.StatusEnum || (Job.StatusEnum = {}));
    })(Job = JobManagerModels.Job || (JobManagerModels.Job = {}));
    /**
     * @export
     * @namespace ScheduleDetails
     */
    let ScheduleDetails;
    (function (ScheduleDetails) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum["RUNNING"] = "RUNNING";
            StatusEnum["STOPPED"] = "STOPPED";
        })(StatusEnum = ScheduleDetails.StatusEnum || (ScheduleDetails.StatusEnum = {}));
    })(ScheduleDetails = JobManagerModels.ScheduleDetails || (JobManagerModels.ScheduleDetails = {}));
})(JobManagerModels = exports.JobManagerModels || (exports.JobManagerModels = {}));
