// @mindconnect/mindconnect-nodejs/dist/src/api/mindconnect-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnlineStatus = exports.DiagnosticInformation = exports.Validity = exports.Mapping = exports.COLLECTION_FORMATS = exports.Order = exports.OnboardingStatus = exports.DataPoint = exports.ClientIdentifier = exports.AgentUpdate = exports.Agent = exports.ConfigurationStyle = void 0;
/**
 * used for Generate Configuration method
 *
 * @export
 * @enum {number}
 */
var ConfigurationStyle;
(function (ConfigurationStyle) {
    ConfigurationStyle["Numeric"] = "NUMERIC";
    ConfigurationStyle["Descriptive"] = "DESCRIPTIVE";
})(ConfigurationStyle = exports.ConfigurationStyle || (exports.ConfigurationStyle = {}));
/**
 * @export
 * @namespace Agent
 */
var Agent;
(function (Agent) {
    /**
     * @export
     * @enum {string}
     */
    let SecurityProfileEnum;
    (function (SecurityProfileEnum) {
        SecurityProfileEnum[SecurityProfileEnum["SHAREDSECRET"] = "SHARED_SECRET"] = "SHAREDSECRET";
        SecurityProfileEnum[SecurityProfileEnum["RSA3072"] = "RSA_3072"] = "RSA3072";
    })(SecurityProfileEnum = Agent.SecurityProfileEnum || (Agent.SecurityProfileEnum = {}));
})(Agent = exports.Agent || (exports.Agent = {}));
/**
 * @export
 * @namespace AgentUpdate
 */
var AgentUpdate;
(function (AgentUpdate) {
    /**
     * @export
     * @enum {string}
     */
    let SecurityProfileEnum;
    (function (SecurityProfileEnum) {
        SecurityProfileEnum[SecurityProfileEnum["SHAREDSECRET"] = "SHARED_SECRET"] = "SHAREDSECRET";
        SecurityProfileEnum[SecurityProfileEnum["RSA3072"] = "RSA_3072"] = "RSA3072";
    })(SecurityProfileEnum = AgentUpdate.SecurityProfileEnum || (AgentUpdate.SecurityProfileEnum = {}));
})(AgentUpdate = exports.AgentUpdate || (exports.AgentUpdate = {}));
/**
 * @export
 * @namespace ClientIdentifier
 */
var ClientIdentifier;
(function (ClientIdentifier) {
    /**
     * @export
     * @enum {string}
     */
    let GrantTypesEnum;
    (function (GrantTypesEnum) {
        GrantTypesEnum[GrantTypesEnum["SHAREDSECRET"] = "SHARED_SECRET"] = "SHAREDSECRET";
        GrantTypesEnum[GrantTypesEnum["RSA3072"] = "RSA_3072"] = "RSA3072";
    })(GrantTypesEnum = ClientIdentifier.GrantTypesEnum || (ClientIdentifier.GrantTypesEnum = {}));
    /**
     * @export
     * @enum {string}
     */
    let TokenEndpointAuthMethodEnum;
    (function (TokenEndpointAuthMethodEnum) {
        TokenEndpointAuthMethodEnum[TokenEndpointAuthMethodEnum["ClientSecretJwt"] = "client_secret_jwt"] = "ClientSecretJwt";
        TokenEndpointAuthMethodEnum[TokenEndpointAuthMethodEnum["PrivateKeyJwt"] = "private_key_jwt"] = "PrivateKeyJwt";
    })(TokenEndpointAuthMethodEnum = ClientIdentifier.TokenEndpointAuthMethodEnum || (ClientIdentifier.TokenEndpointAuthMethodEnum = {}));
})(ClientIdentifier = exports.ClientIdentifier || (exports.ClientIdentifier = {}));
/**
 * @export
 * @namespace DataPoint
 */
var DataPoint;
(function (DataPoint) {
    /**
     * @export
     * @enum {string}
     */
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["INT"] = "INT"] = "INT";
        TypeEnum[TypeEnum["LONG"] = "LONG"] = "LONG";
        TypeEnum[TypeEnum["DOUBLE"] = "DOUBLE"] = "DOUBLE";
        TypeEnum[TypeEnum["BOOLEAN"] = "BOOLEAN"] = "BOOLEAN";
        TypeEnum[TypeEnum["STRING"] = "STRING"] = "STRING";
        TypeEnum[TypeEnum["BIG_STRING"] = "BIG_STRING"] = "BIG_STRING";
        TypeEnum[TypeEnum["TIMESTAMP"] = "TIMESTAMP"] = "TIMESTAMP";
    })(TypeEnum = DataPoint.TypeEnum || (DataPoint.TypeEnum = {}));
})(DataPoint = exports.DataPoint || (exports.DataPoint = {}));
/**
 * @export
 * @namespace OnboardingStatus
 */
var OnboardingStatus;
(function (OnboardingStatus) {
    /**
     * @export
     * @enum {string}
     */
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["NOTONBOARDED"] = "NOT_ONBOARDED"] = "NOTONBOARDED";
        StatusEnum[StatusEnum["ONBOARDING"] = "ONBOARDING"] = "ONBOARDING";
        StatusEnum[StatusEnum["ONBOARDED"] = "ONBOARDED"] = "ONBOARDED";
    })(StatusEnum = OnboardingStatus.StatusEnum || (OnboardingStatus.StatusEnum = {}));
})(OnboardingStatus = exports.OnboardingStatus || (exports.OnboardingStatus = {}));
/**
 * @export
 * @namespace Order
 */
var Order;
(function (Order) {
    /**
     * @export
     * @enum {string}
     */
    let DirectionEnum;
    (function (DirectionEnum) {
        DirectionEnum[DirectionEnum["ASC"] = "ASC"] = "ASC";
        DirectionEnum[DirectionEnum["DESC"] = "DESC"] = "DESC";
    })(DirectionEnum = Order.DirectionEnum || (Order.DirectionEnum = {}));
    /**
     * @export
     * @enum {string}
     */
    let NullHandlingEnum;
    (function (NullHandlingEnum) {
        NullHandlingEnum[NullHandlingEnum["NATIVE"] = "NATIVE"] = "NATIVE";
        NullHandlingEnum[NullHandlingEnum["NULLSFIRST"] = "NULLS_FIRST"] = "NULLSFIRST";
        NullHandlingEnum[NullHandlingEnum["NULLSLAST"] = "NULLS_LAST"] = "NULLSLAST";
    })(NullHandlingEnum = Order.NullHandlingEnum || (Order.NullHandlingEnum = {}));
})(Order = exports.Order || (exports.Order = {}));
/// MindConnect API
/**
 *
 * @export
 */
exports.COLLECTION_FORMATS = {
    csv: ",",
    ssv: " ",
    tsv: "\t",
    pipes: "|",
};
/**
 * @export
 * @namespace Mapping
 */
var Mapping;
(function (Mapping) {
    /**
     * @export
     * @enum {string}
     */
    let DataPointTypeEnum;
    (function (DataPointTypeEnum) {
        DataPointTypeEnum[DataPointTypeEnum["INT"] = "INT"] = "INT";
        DataPointTypeEnum[DataPointTypeEnum["LONG"] = "LONG"] = "LONG";
        DataPointTypeEnum[DataPointTypeEnum["DOUBLE"] = "DOUBLE"] = "DOUBLE";
        DataPointTypeEnum[DataPointTypeEnum["BOOLEAN"] = "BOOLEAN"] = "BOOLEAN";
        DataPointTypeEnum[DataPointTypeEnum["STRING"] = "STRING"] = "STRING";
        DataPointTypeEnum[DataPointTypeEnum["BIGSTRING"] = "BIG_STRING"] = "BIGSTRING";
        DataPointTypeEnum[DataPointTypeEnum["TIMESTAMP"] = "TIMESTAMP"] = "TIMESTAMP";
    })(DataPointTypeEnum = Mapping.DataPointTypeEnum || (Mapping.DataPointTypeEnum = {}));
    /**
     * @export
     * @enum {string}
     */
    let PropertyTypeEnum;
    (function (PropertyTypeEnum) {
        PropertyTypeEnum[PropertyTypeEnum["INT"] = "INT"] = "INT";
        PropertyTypeEnum[PropertyTypeEnum["LONG"] = "LONG"] = "LONG";
        PropertyTypeEnum[PropertyTypeEnum["DOUBLE"] = "DOUBLE"] = "DOUBLE";
        PropertyTypeEnum[PropertyTypeEnum["BOOLEAN"] = "BOOLEAN"] = "BOOLEAN";
        PropertyTypeEnum[PropertyTypeEnum["STRING"] = "STRING"] = "STRING";
        PropertyTypeEnum[PropertyTypeEnum["BIGSTRING"] = "BIG_STRING"] = "BIGSTRING";
        PropertyTypeEnum[PropertyTypeEnum["TIMESTAMP"] = "TIMESTAMP"] = "TIMESTAMP";
    })(PropertyTypeEnum = Mapping.PropertyTypeEnum || (Mapping.PropertyTypeEnum = {}));
})(Mapping = exports.Mapping || (exports.Mapping = {}));
/**
 * @export
 * @namespace Validity
 */
var Validity;
(function (Validity) {
    /**
     * @export
     * @enum {string}
     */
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["VALID"] = "VALID"] = "VALID";
        StatusEnum[StatusEnum["INVALID"] = "INVALID"] = "INVALID";
    })(StatusEnum = Validity.StatusEnum || (Validity.StatusEnum = {}));
    /**
     * @export
     * @enum {string}
     */
    let ReasonsEnum;
    (function (ReasonsEnum) {
        ReasonsEnum[ReasonsEnum["MISSINGDATAPOINT"] = "MISSING_DATAPOINT"] = "MISSINGDATAPOINT";
        ReasonsEnum[ReasonsEnum["MISSINGPROPERTY"] = "MISSING_PROPERTY"] = "MISSINGPROPERTY";
        ReasonsEnum[ReasonsEnum["INVALIDTYPE"] = "INVALID_TYPE"] = "INVALIDTYPE";
        ReasonsEnum[ReasonsEnum["INVALIDUNIT"] = "INVALID_UNIT"] = "INVALIDUNIT";
    })(ReasonsEnum = Validity.ReasonsEnum || (Validity.ReasonsEnum = {}));
})(Validity = exports.Validity || (exports.Validity = {}));
/**
 * @export
 * @namespace DiagnosticInformation
 */
var DiagnosticInformation;
(function (DiagnosticInformation) {
    /**
     * @export
     * @enum {string}
     */
    let SeverityEnum;
    (function (SeverityEnum) {
        SeverityEnum[SeverityEnum["INFO"] = "INFO"] = "INFO";
        SeverityEnum[SeverityEnum["WARN"] = "WARN"] = "WARN";
        SeverityEnum[SeverityEnum["ERROR"] = "ERROR"] = "ERROR";
    })(SeverityEnum = DiagnosticInformation.SeverityEnum || (DiagnosticInformation.SeverityEnum = {}));
})(DiagnosticInformation = exports.DiagnosticInformation || (exports.DiagnosticInformation = {}));
var OnlineStatus;
(function (OnlineStatus) {
    /**
     * @export
     * @enum {string}
     */
    let StatusEnum;
    (function (StatusEnum) {
        StatusEnum[StatusEnum["ONLINE"] = "ONLINE"] = "ONLINE";
        StatusEnum[StatusEnum["OFFLINE"] = "OFFLINE"] = "OFFLINE";
    })(StatusEnum = OnlineStatus.StatusEnum || (OnlineStatus.StatusEnum = {}));
})(OnlineStatus = exports.OnlineStatus || (exports.OnlineStatus = {}));
