// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/commanding/commanding-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandingModels = void 0;
var CommandingModels;
(function (CommandingModels) {
    /**
     *
     * @export
     * @class RequiredError
     * @extends {Error}
     */
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    CommandingModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace Command
     */
    let Command;
    (function (Command) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum[StatusEnum["EXECUTING"] = "EXECUTING"] = "EXECUTING";
            StatusEnum[StatusEnum["EXECUTED"] = "EXECUTED"] = "EXECUTED";
            StatusEnum[StatusEnum["FAILED"] = "FAILED"] = "FAILED";
            StatusEnum[StatusEnum["CANCELED"] = "CANCELED"] = "CANCELED";
        })(StatusEnum = Command.StatusEnum || (Command.StatusEnum = {}));
    })(Command = CommandingModels.Command || (CommandingModels.Command = {}));
    /**
     * @export
     * @namespace CommandListView
     */
    let CommandListView;
    (function (CommandListView) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum[StatusEnum["EXECUTING"] = "EXECUTING"] = "EXECUTING";
            StatusEnum[StatusEnum["EXECUTED"] = "EXECUTED"] = "EXECUTED";
            StatusEnum[StatusEnum["FAILED"] = "FAILED"] = "FAILED";
            StatusEnum[StatusEnum["CANCELED"] = "CANCELED"] = "CANCELED";
        })(StatusEnum = CommandListView.StatusEnum || (CommandListView.StatusEnum = {}));
    })(CommandListView = CommandingModels.CommandListView || (CommandingModels.CommandListView = {}));
    /**
     * @export
     * @namespace Job
     */
    let Job;
    (function (Job) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum[StatusEnum["EXECUTING"] = "EXECUTING"] = "EXECUTING";
            StatusEnum[StatusEnum["EXECUTED"] = "EXECUTED"] = "EXECUTED";
            StatusEnum[StatusEnum["CANCELED"] = "CANCELED"] = "CANCELED";
        })(StatusEnum = Job.StatusEnum || (Job.StatusEnum = {}));
    })(Job = CommandingModels.Job || (CommandingModels.Job = {}));
    /**
     * @export
     * @namespace JobListView
     */
    let JobListView;
    (function (JobListView) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum[StatusEnum["EXECUTING"] = "EXECUTING"] = "EXECUTING";
            StatusEnum[StatusEnum["EXECUTED"] = "EXECUTED"] = "EXECUTED";
            StatusEnum[StatusEnum["CANCELED"] = "CANCELED"] = "CANCELED";
        })(StatusEnum = JobListView.StatusEnum || (JobListView.StatusEnum = {}));
    })(JobListView = CommandingModels.JobListView || (CommandingModels.JobListView = {}));
})(CommandingModels = exports.CommandingModels || (exports.CommandingModels = {}));
