// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/tenant/tenant-management-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TenantManagementModels = void 0;
var TenantManagementModels;
(function (TenantManagementModels) {
    /**
     *
     * @export
     * @class RequiredError
     * @extends {Error}
     */
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    TenantManagementModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace ErrorResponse
     */
    let ErrorResponse;
    (function (ErrorResponse) {
        /**
         * @export
         * @enum {string}
         */
        let ErrorTypeEnum;
        (function (ErrorTypeEnum) {
            ErrorTypeEnum[ErrorTypeEnum["FIELDERROR"] = "FIELD_ERROR"] = "FIELDERROR";
            ErrorTypeEnum[ErrorTypeEnum["UNEXPECTEDERROR"] = "UNEXPECTED_ERROR"] = "UNEXPECTEDERROR";
            ErrorTypeEnum[ErrorTypeEnum["INVALIDSTATE"] = "INVALID_STATE"] = "INVALIDSTATE";
            ErrorTypeEnum[ErrorTypeEnum["PARSINGERROR"] = "PARSING_ERROR"] = "PARSINGERROR";
        })(ErrorTypeEnum = ErrorResponse.ErrorTypeEnum || (ErrorResponse.ErrorTypeEnum = {}));
    })(ErrorResponse = TenantManagementModels.ErrorResponse || (TenantManagementModels.ErrorResponse = {}));
    /**
     * @export
     * @namespace FieldError
     */
    let FieldError;
    (function (FieldError) {
        /**
         * @export
         * @enum {string}
         */
        let ErrorTypeEnum;
        (function (ErrorTypeEnum) {
            ErrorTypeEnum[ErrorTypeEnum["INVALIDINPUT"] = "INVALID_INPUT"] = "INVALIDINPUT";
            ErrorTypeEnum[ErrorTypeEnum["VALIDATIONERROR"] = "VALIDATION_ERROR"] = "VALIDATIONERROR";
            ErrorTypeEnum[ErrorTypeEnum["NOTALLOWED"] = "NOT_ALLOWED"] = "NOTALLOWED";
            ErrorTypeEnum[ErrorTypeEnum["UNIQUECONSTRAINTERROR"] = "UNIQUE_CONSTRAINT_ERROR"] = "UNIQUECONSTRAINTERROR";
        })(ErrorTypeEnum = FieldError.ErrorTypeEnum || (FieldError.ErrorTypeEnum = {}));
    })(FieldError = TenantManagementModels.FieldError || (TenantManagementModels.FieldError = {}));
    /**
     * @export
     * @namespace LinkCollection
     */
    let LinkCollection;
    (function (LinkCollection) {
        /**
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum[TypeEnum["Www"] = "www"] = "Www";
            TypeEnum[TypeEnum["Phone"] = "phone"] = "Phone";
            TypeEnum[TypeEnum["Mail"] = "mail"] = "Mail";
        })(TypeEnum = LinkCollection.TypeEnum || (LinkCollection.TypeEnum = {}));
    })(LinkCollection = TenantManagementModels.LinkCollection || (TenantManagementModels.LinkCollection = {}));
    /**
     * @export
     * @namespace TenantInfo
     */
    let TenantInfo;
    (function (TenantInfo) {
        /**
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum[TypeEnum["DEVELOPER"] = "DEVELOPER"] = "DEVELOPER";
            TypeEnum[TypeEnum["OPERATOR"] = "OPERATOR"] = "OPERATOR";
            TypeEnum[TypeEnum["USER"] = "USER"] = "USER";
            TypeEnum[TypeEnum["CUSTOMER"] = "CUSTOMER"] = "CUSTOMER";
        })(TypeEnum = TenantInfo.TypeEnum || (TenantInfo.TypeEnum = {}));
    })(TenantInfo = TenantManagementModels.TenantInfo || (TenantManagementModels.TenantInfo = {}));
})(TenantManagementModels = exports.TenantManagementModels || (exports.TenantManagementModels = {}));
