// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/signal-validation/signal-validation-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignalValidationModels = void 0;
var SignalValidationModels;
(function (SignalValidationModels) {
    /**
     * @export
     * @namespace Bias
     */
    let Bias;
    (function (Bias) {
        /**
         * @export
         * @enum {string}
         */
        let DescriptionEnum;
        (function (DescriptionEnum) {
            DescriptionEnum[DescriptionEnum["BIAS"] = "BIAS"] = "BIAS";
        })(DescriptionEnum = Bias.DescriptionEnum || (Bias.DescriptionEnum = {}));
    })(Bias = SignalValidationModels.Bias || (SignalValidationModels.Bias = {}));
    /**
     * @export
     * @namespace Event
     */
    let Event;
    (function (Event) {
        /**
         * @export
         * @enum {string}
         */
        let DescriptionEnum;
        (function (DescriptionEnum) {
            DescriptionEnum[DescriptionEnum["START"] = "EVENT_START"] = "START";
            DescriptionEnum[DescriptionEnum["END"] = "EVENT_END"] = "END";
        })(DescriptionEnum = Event.DescriptionEnum || (Event.DescriptionEnum = {}));
    })(Event = SignalValidationModels.Event || (SignalValidationModels.Event = {}));
    /**
     * @export
     * @namespace Jump
     */
    let Jump;
    (function (Jump) {
        /**
         * @export
         * @enum {string}
         */
        let DescriptionEnum;
        (function (DescriptionEnum) {
            DescriptionEnum[DescriptionEnum["JUMP"] = "JUMP"] = "JUMP";
        })(DescriptionEnum = Jump.DescriptionEnum || (Jump.DescriptionEnum = {}));
    })(Jump = SignalValidationModels.Jump || (SignalValidationModels.Jump = {}));
    /**
     * @export
     * @namespace Noise
     */
    let Noise;
    (function (Noise) {
        /**
         * @export
         * @enum {string}
         */
        let DescriptionEnum;
        (function (DescriptionEnum) {
            DescriptionEnum[DescriptionEnum["NOISE"] = "NOISE"] = "NOISE";
        })(DescriptionEnum = Noise.DescriptionEnum || (Noise.DescriptionEnum = {}));
    })(Noise = SignalValidationModels.Noise || (SignalValidationModels.Noise = {}));
    /**
     * @export
     * @namespace Range
     */
    let Range;
    (function (Range) {
        /**
         * @export
         * @enum {string}
         */
        let KindEnum;
        (function (KindEnum) {
            KindEnum[KindEnum["UPPERLIMIT"] = "EVENT_UPPER_LIMIT"] = "UPPERLIMIT";
            KindEnum[KindEnum["LOWERLIMIT"] = "EVENT_LOWER_LIMIT"] = "LOWERLIMIT";
        })(KindEnum = Range.KindEnum || (Range.KindEnum = {}));
    })(Range = SignalValidationModels.Range || (SignalValidationModels.Range = {}));
    /**
     * @export
     * @namespace Spike
     */
    let Spike;
    (function (Spike) {
        /**
         * @export
         * @enum {string}
         */
        let DescriptionEnum;
        (function (DescriptionEnum) {
            DescriptionEnum[DescriptionEnum["SPIKE"] = "SPIKE"] = "SPIKE";
        })(DescriptionEnum = Spike.DescriptionEnum || (Spike.DescriptionEnum = {}));
    })(Spike = SignalValidationModels.Spike || (SignalValidationModels.Spike = {}));
})(SignalValidationModels = exports.SignalValidationModels || (exports.SignalValidationModels = {}));
