// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/utc/utc-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsageTransparencyModels = void 0;
var UsageTransparencyModels;
(function (UsageTransparencyModels) {
    /**
     *
     * @export
     * @class RequiredError
     * @extends {Error}
     */
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    UsageTransparencyModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace UsageError
     */
    let UsageError;
    (function (UsageError) {
        /**
         * @export
         * @enum {string}
         */
        let ErrorTypeEnum;
        (function (ErrorTypeEnum) {
            ErrorTypeEnum[ErrorTypeEnum["ValidationError"] = "ValidationError"] = "ValidationError";
            ErrorTypeEnum[ErrorTypeEnum["BillingError"] = "BillingError"] = "BillingError";
            ErrorTypeEnum[ErrorTypeEnum["OtherError"] = "OtherError"] = "OtherError";
        })(ErrorTypeEnum = UsageError.ErrorTypeEnum || (UsageError.ErrorTypeEnum = {}));
    })(UsageError = UsageTransparencyModels.UsageError || (UsageTransparencyModels.UsageError = {}));
    /**
     * @export
     * @namespace UsagesJob
     */
    let UsagesJob;
    (function (UsagesJob) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum[StatusEnum["ACCEPTED"] = "ACCEPTED"] = "ACCEPTED";
            StatusEnum[StatusEnum["INPROGRESS"] = "INPROGRESS"] = "INPROGRESS";
            StatusEnum[StatusEnum["ERRORS"] = "ERRORS"] = "ERRORS";
            StatusEnum[StatusEnum["COMPLETED"] = "COMPLETED"] = "COMPLETED";
        })(StatusEnum = UsagesJob.StatusEnum || (UsagesJob.StatusEnum = {}));
    })(UsagesJob = UsageTransparencyModels.UsagesJob || (UsageTransparencyModels.UsagesJob = {}));
    /**
     * @export
     * @namespace UsagesSummary
     */
    let UsagesSummary;
    (function (UsagesSummary) {
        /**
         * @export
         * @enum {string}
         */
        let ProcessStatusEnum;
        (function (ProcessStatusEnum) {
            ProcessStatusEnum[ProcessStatusEnum["ACCEPTED"] = "ACCEPTED"] = "ACCEPTED";
            ProcessStatusEnum[ProcessStatusEnum["VERIFIED"] = "VERIFIED"] = "VERIFIED";
            ProcessStatusEnum[ProcessStatusEnum["VERIFICATIONFAILED"] = "VERIFICATIONFAILED"] = "VERIFICATIONFAILED";
            ProcessStatusEnum[ProcessStatusEnum["NOVERIFICATION"] = "NOVERIFICATION"] = "NOVERIFICATION";
            ProcessStatusEnum[ProcessStatusEnum["AGGREGATED"] = "AGGREGATED"] = "AGGREGATED";
            ProcessStatusEnum[ProcessStatusEnum["SENTFORBILLING"] = "SENTFORBILLING"] = "SENTFORBILLING";
            ProcessStatusEnum[ProcessStatusEnum["FAILEDTOSENDFORBILLING"] = "FAILEDTOSENDFORBILLING"] = "FAILEDTOSENDFORBILLING";
            ProcessStatusEnum[ProcessStatusEnum["RECEIVEDFORBILLING"] = "RECEIVEDFORBILLING"] = "RECEIVEDFORBILLING";
        })(ProcessStatusEnum = UsagesSummary.ProcessStatusEnum || (UsagesSummary.ProcessStatusEnum = {}));
    })(UsagesSummary = UsageTransparencyModels.UsagesSummary || (UsageTransparencyModels.UsagesSummary = {}));
    /**
     * @export
     * @namespace User
     */
    let User;
    (function (User) {
        /**
         * @export
         * @enum {string}
         */
        let UserTypeEnum;
        (function (UserTypeEnum) {
            UserTypeEnum["Tenant"] = "tenant";
            UserTypeEnum["User"] = "user";
            UserTypeEnum["Agent"] = "agent";
        })(UserTypeEnum = User.UserTypeEnum || (User.UserTypeEnum = {}));
    })(User = UsageTransparencyModels.User || (UsageTransparencyModels.User = {}));
})(UsageTransparencyModels = exports.UsageTransparencyModels || (exports.UsageTransparencyModels = {}));
