// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/notification-v4/notification-v4-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationModelsV4 = void 0;
var NotificationModelsV4;
(function (NotificationModelsV4) {
    /**
     *
     * @export
     * @class RequiredError
     * @extends {Error}
     */
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    NotificationModelsV4.RequiredError = RequiredError;
    /**
     * @export
     * @namespace AppRegistrationRequest
     */
    let AppRegistrationRequest;
    (function (AppRegistrationRequest) {
        /**
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum["Ios"] = "ios";
            TypeEnum["Android"] = "android";
        })(TypeEnum = AppRegistrationRequest.TypeEnum || (AppRegistrationRequest.TypeEnum = {}));
    })(AppRegistrationRequest = NotificationModelsV4.AppRegistrationRequest || (NotificationModelsV4.AppRegistrationRequest = {}));
    /**
     * @export
     * @namespace AppRegistrationResponse
     */
    let AppRegistrationResponse;
    (function (AppRegistrationResponse) {
        /**
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum["Ios"] = "ios";
            TypeEnum["Android"] = "android";
        })(TypeEnum = AppRegistrationResponse.TypeEnum || (AppRegistrationResponse.TypeEnum = {}));
    })(AppRegistrationResponse = NotificationModelsV4.AppRegistrationResponse || (NotificationModelsV4.AppRegistrationResponse = {}));
    /**
     * @export
     * @namespace DispatchStatus
     */
    let DispatchStatus;
    (function (DispatchStatus) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum["Dispatched"] = "dispatched";
            StatusEnum["Failed"] = "failed";
        })(StatusEnum = DispatchStatus.StatusEnum || (DispatchStatus.StatusEnum = {}));
    })(DispatchStatus = NotificationModelsV4.DispatchStatus || (NotificationModelsV4.DispatchStatus = {}));
    /**
     * @export
     * @namespace DispatchStatusSMS
     */
    let DispatchStatusSMS;
    (function (DispatchStatusSMS) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum["Dispatched"] = "dispatched";
            StatusEnum["Failed"] = "failed";
        })(StatusEnum = DispatchStatusSMS.StatusEnum || (DispatchStatusSMS.StatusEnum = {}));
    })(DispatchStatusSMS = NotificationModelsV4.DispatchStatusSMS || (NotificationModelsV4.DispatchStatusSMS = {}));
    /**
     * @export
     * @namespace MobileAppObject
     */
    let MobileAppObject;
    (function (MobileAppObject) {
        /**
         * ! fix: (April 2021) Manual fix to correspond with Implementation (iOS instead of ios)
         * @export
         * @enum {string}
         */
        let TypeEnum;
        (function (TypeEnum) {
            TypeEnum["Ios"] = "iOS";
            TypeEnum["Android"] = "android";
        })(TypeEnum = MobileAppObject.TypeEnum || (MobileAppObject.TypeEnum = {}));
    })(MobileAppObject = NotificationModelsV4.MobileAppObject || (NotificationModelsV4.MobileAppObject = {}));
    /**
     * @export
     * @namespace MulticastEmailNotificationJob
     */
    let MulticastEmailNotificationJob;
    (function (MulticastEmailNotificationJob) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum["Queued"] = "queued";
            StatusEnum["Dispatched"] = "dispatched";
            StatusEnum["DispatchedWithFailures"] = "dispatchedWithFailures";
            StatusEnum["Failed"] = "failed";
        })(StatusEnum = MulticastEmailNotificationJob.StatusEnum || (MulticastEmailNotificationJob.StatusEnum = {}));
    })(MulticastEmailNotificationJob = NotificationModelsV4.MulticastEmailNotificationJob || (NotificationModelsV4.MulticastEmailNotificationJob = {}));
    /**
     * @export
     * @namespace MulticastEmailNotificationRequestMetadata
     */
    let MulticastEmailNotificationRequestMetadata;
    (function (MulticastEmailNotificationRequestMetadata) {
        /**
         * @export
         * @enum {string}
         */
        let PriorityEnum;
        (function (PriorityEnum) {
            PriorityEnum["Low"] = "Low";
            PriorityEnum["Normal"] = "Normal";
            PriorityEnum["High"] = "High";
        })(PriorityEnum = MulticastEmailNotificationRequestMetadata.PriorityEnum || (MulticastEmailNotificationRequestMetadata.PriorityEnum = {}));
    })(MulticastEmailNotificationRequestMetadata = NotificationModelsV4.MulticastEmailNotificationRequestMetadata || (NotificationModelsV4.MulticastEmailNotificationRequestMetadata = {}));
    /**
     * @export
     * @namespace MulticastSMSNotificationJob
     */
    let MulticastSMSNotificationJob;
    (function (MulticastSMSNotificationJob) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum["Queued"] = "queued";
            StatusEnum["Dispatched"] = "dispatched";
            StatusEnum["Failed"] = "failed";
        })(StatusEnum = MulticastSMSNotificationJob.StatusEnum || (MulticastSMSNotificationJob.StatusEnum = {}));
    })(MulticastSMSNotificationJob = NotificationModelsV4.MulticastSMSNotificationJob || (NotificationModelsV4.MulticastSMSNotificationJob = {}));
    /**
     * @export
     * @namespace SendResponse
     */
    let SendResponse;
    (function (SendResponse) {
        /**
         * @export
         * @enum {string}
         */
        let StatusEnum;
        (function (StatusEnum) {
            StatusEnum["Queued"] = "queued";
            StatusEnum["Dispatched"] = "dispatched";
            StatusEnum["Failed"] = "failed";
        })(StatusEnum = SendResponse.StatusEnum || (SendResponse.StatusEnum = {}));
    })(SendResponse = NotificationModelsV4.SendResponse || (NotificationModelsV4.SendResponse = {}));
})(NotificationModelsV4 = exports.NotificationModelsV4 || (exports.NotificationModelsV4 = {}));
