// @mindconnect/mindconnect-nodejs/dist/src/api/sdk/spectrum/spectrum-analysis-models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpectrumAnalysisModels = void 0;
var SpectrumAnalysisModels;
(function (SpectrumAnalysisModels) {
    /**
     *
     * @export
     * @class RequiredError
     * @extends {Error}
     */
    class RequiredError extends Error {
        constructor(field, msg) {
            super(msg);
            this.field = field;
            this.name = "RequiredError";
        }
    }
    SpectrumAnalysisModels.RequiredError = RequiredError;
    /**
     * @export
     * @namespace WindowType
     */
    let WindowType;
    (function (WindowType) {
        /**
         * @export
         * @enum {string}
         */
        let WindowTypeEnum;
        (function (WindowTypeEnum) {
            WindowTypeEnum[WindowTypeEnum["FLATTOP"] = "FLATTOP"] = "FLATTOP";
            WindowTypeEnum[WindowTypeEnum["HAMMING"] = "HAMMING"] = "HAMMING";
            WindowTypeEnum[WindowTypeEnum["HANNING"] = "HANNING"] = "HANNING";
            WindowTypeEnum[WindowTypeEnum["BLACKMAN"] = "BLACKMAN"] = "BLACKMAN";
        })(WindowTypeEnum = WindowType.WindowTypeEnum || (WindowType.WindowTypeEnum = {}));
    })(WindowType = SpectrumAnalysisModels.WindowType || (SpectrumAnalysisModels.WindowType = {}));
})(SpectrumAnalysisModels = exports.SpectrumAnalysisModels || (exports.SpectrumAnalysisModels = {}));
